### PR TITLE
[CLIP]: Sparseml onnx export pathways for open_clip models

### PR DESCRIPTION
**Pull request by dsikka**
### Summary:

- The current onnx export functionality using OpenClip requires torch nightly and opset = 14
- This PR provides an example of how to use sparseml's onnx export functionality to export the visual and text branches of the CLIP model, using [OpenClip](https://github.com/mlfoundations/open_clip)
- OpenClip wraps everything in a `torch.nn.Module` object for the visual branch (ie everything required for the forward pass) but does not do this for the text branch. For the sake of completeness, the `TextModel` class was added (see `clip_models.py`) which wraps the appropriate steps needed in the forward pass. This is then exported 
- Furthermore, for captioning, OpenClip provides thee CoCa models, which requires an additional text-decoder and also uses a separate interface compared to the other models. As such, there is a special `is_coca` case provided in the scripts

### Test Plan:
- Ran this locally (requires openclip and torch nightly to be installed) and created a `convnext_onnx` directory with the `clip_text.onnx` and `clip_visual.onnx` exports
```
 python clip_onnx_export.py --model convnext_base_w_320 --pretrained laion_aesthetic_s13b_b82k --export-path convnext_onnx
```

- Also ran a CoCa-specific command which created the default `clip_onnx` directory with the visual and text exports, along with the text decoder export `clip_text_decoder.onnx`
```
python clip_onnx_export.py --model coca_ViT-L-14 --pretrained laion2b_s13b_b90k
```

### CODE REVIEW

File: integrations/clip/clip_models.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+    def forward(self, input_ids):
+        x = self.token_embedding(input_ids).to(self.cast_dtype)
+        x = x + self.positional_embedding.to(self.cast_dtype)
+        x = x.permute(1, 0, 2)  # NLD -> LND
+        x = self.transformer(x, attn_mask=self.attn_mask)
+        x = x.permute(1, 0, 2)  # LND -> NLD
+        x = self.ln_final(x)  # [batch_size, n_ctx, transformer.width]
+        # take features from the eot embedding (eot_token = highest in each sequence)
+        x = x[torch.arange(x.shape[0]), input_ids.argmax(dim=-1)] @ self.text_projection
```
Comment (bfineran) [inline integrations/clip/clip_models.py:63]: what semantically is the output here? are we only returning scores for the highest activations?

File: integrations/clip/clip_models.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        transformer,
+        ln_final,
+        text_projection,
+        attn_mask,
+    ):
+
+        super().__init__()
+
+        self.token_embedding = token_embedding
+        self.tokenizer = tokenizer
```
Comment (bfineran) [inline integrations/clip/clip_models.py:47]: where is this tokenizer used?

Comment (dsikka) [reply integrations/clip/clip_models.py:47]: Just for the sample batch. Added to the `TextModel` object for the sake of completeness but not needed in the forward pass. Can remove

File: integrations/clip/clip_models.py (hunk 1 above)
Comment (dsikka) [reply integrations/clip/clip_models.py:63]: In the last time, they seems to only take features from the dimension with the end-of-text token, reducing the dimension from `batch x context_length x transformer.width` to `batch x transformer.width` which is then projected

File: integrations/clip/clip_models.py (hunk 2 above)
Comment (bfineran) [reply integrations/clip/clip_models.py:47]: makes sense, let's slim down any args not needed for onnx export / forward pass

**Review by bfineran — APPROVED** [review]
LGTM pending comments

**Review by bfineran — DISMISSED** [review]

File: integrations/clip/clip_onnx_export.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+def _export_text(
+    model: torch.nn.Module,
+    device: str,
+    export_path: Union[str, Path],
+    tokenizer,
+    is_coca: bool,
+    **export_kwargs,
+):
+    module_name = "clip_text.onnx"
```
Comment (rahul-tuli) [inline integrations/clip/clip_onnx_export.py:86]: Would this fail or overwrite the model if existing `clip_text.onnx` is present? Same comment for other hardcoded names

File: integrations/clip/clip_onnx_export.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+"""
+Examples on how to use sparsemls's onnx export functionality to export CLIP visual
+and text models using the OpenCLIP API.
+
+Note: This requires torch nightly and openclip to be installed:
+https://github.com/mlfoundations/open_clip
+
+"""
+import argparse
```
[the remark this answers is not in the document]
Comment (dsikka) [reply integrations/clip/clip_onnx_export.py:23]: Is this necessary for an example? I'd rather not introduce it at this point in the dev cycle. Potentially in a follow-up PR?

**Review by KSGulin — APPROVED** [review]

File: integrations/clip/clip_onnx_export.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rahul-tuli) [reply integrations/clip/clip_onnx_export.py:23]: Sure, works for me as long as we do scope it in.!

File: integrations/clip/clip_onnx_export.py (hunk 3 above)
Comment (dsikka) [reply integrations/clip/clip_onnx_export.py:86]: It won't fail. Will just write over the existing files.

**Review by rahul-tuli — APPROVED** [review]
LGTM!

**Review by KSGulin — APPROVED** [review]

**Pull request merged**